### PR TITLE
[Refactor] Deduplicate final commitment validity checks

### PR DESCRIPTION
**Pull request by random-zebra**
And remove circular dependency `evo/deterministicmns` <-> `llmq/quorums_utils`.

### CODE REVIEW

**Comment by furszy** [comment]
rebase needed.

**Comment by random-zebra** [comment]
Rebased

File: src/evo/specialtx_validation.cpp [hunk 1]
```diff
@@ -420,57 +420,79 @@ static bool CheckProUpRevTx(const CTransaction& tx, const CBlockIndex* pindexPre
...
+        // Check height
+        if (pindexQuorum->nHeight % params->dkgInterval != 0) {
+            // not first block of DKG interval
+            return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-height");
+        }
+
+        // Get members and check signatures (for not-null commitments)
+        std::vector<CBLSPublicKey> allkeys;
+        if (!qfc.IsNull()) {
+            for (const auto m : deterministicMNManager->GetAllQuorumMembers((Consensus::LLMQType)qfc.llmqType, pindexQuorum)) {
```
Comment (furszy) [inline src/evo/specialtx_validation.cpp:464]: ref -> `const auto& m`

File: src/evo/specialtx_validation.cpp [hunk 2]
```diff
@@ -420,57 +420,79 @@ static bool CheckProUpRevTx(const CTransaction& tx, const CBlockIndex* pindexPre
...
+        // Get quorum index
+        auto it = mapBlockIndex.find(qfc.quorumHash);
+        if (it == mapBlockIndex.end()) {
             return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-hash");
         }
-        const CBlockIndex* pindexQuorum = mapBlockIndex.at(pl.commitment.quorumHash);
+        const CBlockIndex* pindexQuorum = it->second;
         if (pindexQuorum != pindexPrev->GetAncestor(pindexQuorum->nHeight)) {
             // not part of active chain
             return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-hash");
```
Comment (furszy) [inline src/evo/specialtx_validation.cpp:452]: Let's have another error code for this DoS so we can test both invalid paths.
One is from an inexistent block while the other one is from a block that is not part of the main chain.

File: src/evo/specialtx_validation.cpp [hunk 3]
```diff
@@ -420,57 +420,79 @@ static bool CheckProUpRevTx(const CTransaction& tx, const CBlockIndex* pindexPre
...
-            return state.DoS(100, false, REJECT_INVALID, "bad-qc-invalid");
+        // Check height
+        if (pindexQuorum->nHeight % params->dkgInterval != 0) {
+            // not first block of DKG interval
+            return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-height");
+        }
+
+        // Get members and check signatures (for not-null commitments)
+        std::vector<CBLSPublicKey> allkeys;
+        if (!qfc.IsNull()) {
```
Comment (furszy) [inline src/evo/specialtx_validation.cpp:463]: Can move `allkeys` inside the not null qfc block.

**Review by furszy — APPROVED** [review]
ACK 2dcbe325011ab443d48b7d74a343b1684bf4e0fb

Would be good to add a 7c) test case for a qfc with a quorum hash from a fork (here or in a follow up work).

File: src/evo/specialtx_validation.cpp [hunk 4]
```diff
@@ -432,12 +432,12 @@ static bool CheckLLMQCommitment(const CTransaction& tx, const CBlockIndex* pinde
         return state.DoS(100, false, REJECT_INVALID, "bad-qc-version");
     }
 
-    if (!Params().GetConsensus().llmqs.count((Consensus::LLMQType)pl.commitment.llmqType)) {
+    Optional<Consensus::LLMQParams> params = Params().GetConsensus().GetLLMQParams(pl.commitment.llmqType);
+    if (params == nullopt) {
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/evo/specialtx_validation.cpp:434]: yea. I prefer to make it explicit.

File: src/evo/specialtx_validation.cpp (hunk 2 above)
Comment (random-zebra) [reply src/evo/specialtx_validation.cpp:452]: Sure, but better to add it (with its test) in a small follow-up PR.

File: src/evo/specialtx_validation.cpp (hunk 3 above)
Comment (random-zebra) [reply src/evo/specialtx_validation.cpp:463]: Yep. Done.

**Review by furszy — DISMISSED** [review]
utACK aa03eac0 after feedback squash.

**Review by furszy — APPROVED** [review]
ACK 9e35c74

**Review by Fuzzbawls — APPROVED** [review]
ACK 9e35c740a792496fd9f5df8ba31c9d554000de78

**Pull request merged**